Let `SceneFileWriter` access `ffmpeg` via `av` instead of via external process

behackl opened the pull request:
Potentially, this is a pretty substantial improvement to the overall usability of Manim:

- `av` provides wheels for *a lot* of architectures that come with a statically linked version of `ffmpeg` (-> users would no longer need to install it on their own)
- As a result, we could also throw all of the extra steps in our CI pipelines that setup ffmpeg away
- Plus documentation build times should increase too as less setup is needed; but not only that, ...
- Piping to an external process is potentially slower than inner-process communication; in a non-exact measurement on my local device I found `OpeningManim` in `example_scenes/basic.py` to require only around 7.7sec CPU time on the current `main` branch vs. 5.7sec CPU time on this branch. (Unsure whether this is significant, we'll see in the documentation build times.)

Depending on how experiments go, there is a lot more to do here:
- [x] Refactor partial movie file creation
- [x] Figure out how video filters work exactly (we currently need `-vf vflip` for OpenGL rendering ... or we solve that differently)
- [x] Refactor combination of partial movie files into final output file
- [x] Test creation of movies with other codecs

The following tasks will be tackled in follow-up PRs (note that actually, the documentation has already been adapted and mentions of `ffmpeg` have been removed -- however, as Manim is effectively just `pip`-installable with this change, we should aggressively simplify the installation instructions):
- Make sure users can also interact with the av settings somehow
- Simplify installation documentation (simply `pip install manim`, plus optional LaTeX installation ... 👀)

jeertmans commented on manim/scene/scene_file_writer.py:
```diff
@@ -388,15 +381,17 @@ def write_frame(self, frame_or_renderer: np.ndarray | OpenGLRenderer):
-            self.writing_process.stdin.write(
-                renderer.get_raw_frame_buffer_object_data(),
-            )
+            av_frame = av.VideoFrame.from_ndarray(renderer.get_frame(), format="rgba")
```
I think we can apply the `-vf vlip` filter directly on the ndarray object, since it just flips one axis.

jeertmans replied:
After checking, seems like `get_frame()` already flips the vertical axis with `np.flipud`:

https://github.com/ManimCommunity/manim/blob/40766cb648de05941694bacb773812903db2a817/manim/renderer/opengl_renderer.py#L559-L566

jeertmans commented on pyproject.toml:
```diff
@@ -54,6 +54,7 @@ svgelements = ">=1.8.0"
 tqdm = ">=4.0.0"
 typing-extensions = ">=4.0.0"
 watchdog = ">=2.0.0"
+av = "^11.0.0"
```
Put this at the top of the deps (after `python`) so we keep them ordered?

behackl replied:
Yes, and probably also make the version restriction less strict (cc @Viicos)

jeertmans replied:
Regarding the "less strict" Python version, I think this is a general issue with Poetry where if a dependency you need requires some Python version, you then have to explicitly specify a Python constraint.

Viicos replied:
Yes `av>=some_version` should be ok. Looking at the [changelog](https://github.com/PyAV-Org/PyAV/blob/main/CHANGELOG.rst), it seems pretty stable so we can probably go for `>=9.0.0` or something? This can be bumped of course if you need some features in higher versions.

> Regarding the "less strict" Python version, I think this is a general issue with Poetry where if a dependency you need requires some Python version, you then have to explicitly specify a Python constraint.

I think @behackl was talking about the constraint on `av`. But yeah regarding the Python version we unfortunately had to add a higher bound

jeertmans replied:
@Viicos got it :p

jeertmans commented on manim/scene/scene_file_writer.py:
```diff
@@ -388,15 +381,17 @@ def write_frame(self, frame_or_renderer: np.ndarray | OpenGLRenderer):
             frame = frame_or_renderer
             if write_to_movie():
-                self.writing_process.stdin.write(frame.tobytes())
+                av_frame = av.VideoFrame.from_ndarray(frame, format="rgba")
+                for packet in self.video_stream.encode(av_frame):
+                    self.video_container.mux(packet)
             if is_png_format() and not config["dry_run"]:
                 self.output_image_from_array(frame)
 
     def write_opengl_frame(self, renderer: OpenGLRenderer):
         if write_to_movie():
-            self.writing_process.stdin.write(
-                renderer.get_raw_frame_buffer_object_data(),
-            )
+            av_frame = av.VideoFrame.from_ndarray(renderer.get_frame(), format="rgba")
```
As I mentioned in another comment, I think we should take the `n_frames: int = 1` option to this method and add the same multiple times **after** we converted it into a `av.VideoFrame` and got the packets (this is a list, so it does not exhaust). I did not measure any significant speed-up yet, but it does not make sense to me to do the conversion multiple times when you could avoid it very easily.

jeertmans commented on manim/mobject/geometry/polygram.py:
```diff
@@ -472,7 +472,6 @@ class Star(Polygon):
     Examples
     --------
     .. manim:: StarExample
-        :save_as_gif:
```
Gif not working anymore?

behackl replied:
The documentation build indeed failed on this example (and thus likely is broken in general, we need to investigate rendering with `--format=gif`), but I also see no good reason for this singular example being rendered as a gif while every other example in our docs is a video (or still image).

jeertmans replied:
I tried to fix GIF but still have some issue:

![image](https://github.com/ManimCommunity/manim/assets/27275099/6da30d89-d090-4e63-b4a5-a02e87b5439f)

(from `output_container.mux`).

jeertmans commented on manim/_config/utils.py:
```diff
@@ -24,6 +24,7 @@
 from pathlib import Path
 from typing import Any, ClassVar, Iterable, Iterator, NoReturn
 
+import av
```
I guess this is needed to setup the `libav` logging? If so, this should be documented and maybe but some `# noqa: <code>` after?

jeertmans replied:
nvm: I don't think importing `av` should be needed

behackl replied:
I actually have not tested whether setting the ffmpeg verbosity really works; the import is from a previous approach (before I read in the `av` docs which method is preferred). I guess it should work without the import too, but haven't tested it yet.

jeertmans replied:
I tried, but I cannot make the `ffmeg_loglevel` work.

I manually set `av.logging.set_level(5)`, which does not output anything too.

jeertmans reviewed (commented):
Nice work, I've left some comments :-)

I'd gladly join forces on that, but could we share a branch where we can both edit on?

Or I can copy your branch, and you can manually rebase yours onto mine from time to time.

MrDiver commented:
Love the work you're doing here! ❤️

jeertmans commented on manim/scene/scene_file_writer.py:
```diff
@@ -567,37 +559,46 @@ def combine_files(
-            ]
-        else:
-            commands += ["-c", "copy"]
+            av_options["vf"] = (  # add video filter (is there a better way to do this?)
+                f"fps={np.clip(config['frame_rate'], 1, 50)},"
+                "split[s0][s1];[s0]palettegen=stats_mode=diff[p];"
+                "[s1][p]paletteuse=dither=bayer:bayer_scale=5:diff_mode=rectangle"
+            )
```
Seems like video filters must be applied separately: https://github.com/PyAV-Org/PyAV/issues/239

behackl replied:
Yes, I've tried the `Graph`-construction that you have to push/pull frames through initially, but was happy when you came up with the alternative `options`-approach.

Well, seems like we'll have to look into it regardless.

jeertmans replied:
But does the `options` approach work? I'm trying to move Manim Slides to use `av` too, and cannot make the reverse filter work.

jeertmans commented:
@behackl any idea why tests might fail in CI but not on my computer?

![image](https://github.com/ManimCommunity/manim/assets/27275099/62fe8312-4169-482f-b1fc-3901dbbd0346)

behackl commented:
@jeertmans locally you are probably on an older version of cairo, which we disabled running graphical tests for (check the output, there are ~200 skipped tests).

The tests currently seem to fail because of a mismatch in the number of generated frames in the graphical tests. Curious!

jeertmans commented:
Ah yes... indeed. But https://github.com/ManimCommunity/manim/blob/main/.github/scripts/ci_build_cairo.py install Cairo 1.18, which is not even available from `apt` (Ubuntu 22.04). The latest version available is 1.16.

jeertmans commented:
> Ah yes... indeed. But https://github.com/ManimCommunity/manim/blob/main/.github/scripts/ci_build_cairo.py install Cairo 1.18, which is not even available from `apt` (Ubuntu 22.04). The latest version available is 1.16.

I have raised an issue about that (#3521), as I think this may be good to document this.

> The tests currently seem to fail because of a mismatch in the number of generated frames in the graphical tests. Curious!

Yeah I don't know why it occurs... I have updated the reference frames (not sure if this is the right solution) in https://github.com/ManimCommunity/manim/pull/3501/commits/904cfb46ae8db7dfac01036671991cf81a7823de.

I tested the new manim in a large project I did, and encountered some rendering issue, so will reach back to you when I (hopefully) fixex them!

behackl commented:
I'll investigate this locally.

behackl commented:
Rendering for non-standard output formats needs some work; I believe that the `-c copy` flag currently is just getting ignored. The `.mov` generated when rendering with `-t` appears to be broken.

jeertmans commented:
@behackl maybe that might be good to save video files in various format with the release version of Manim, and compare the output video files with this new version (this is assuming that everything was done correctly in prior versions of Manim)

behackl commented:
> @behackl maybe that might be good to save video files in various format with the release version of Manim, and compare the output video files with this new version (this is assuming that everything was done correctly in prior versions of Manim)

Indeed, I am not even sure in how far correctness of generated files is tested currently. I am thinking of adding some video tests (using the `pyav`-equivalent of `ffprobe`) to check that things like the generated format + codec etc. is actually correct. I'm on this! :+1:

jeertmans commented:
> > @behackl maybe that might be good to save video files in various format with the release version of Manim, and compare the output video files with this new version (this is assuming that everything was done correctly in prior versions of Manim)
> 
> Indeed, I am not even sure in how far correctness of generated files is tested currently. I am thinking of adding some video tests (using the `pyav`-equivalent of `ffprobe`) to check that things like the generated format + codec etc. is actually correct. I'm on this! 👍

Nice, perfect! Tell me if I can help (I don't have much time, but I'd love to have the release :p)

behackl commented:
I might have some time in the next few days to finish this, finally. To do so, I propose the following:

- The task that remains here is the creation of an (ideally parameterized) test to cover all possible options affecting the renderer (mainly with respect to the different codecs and transparency). There are already a bunch of individual tests, but I am not sure that all possible options are covered. I'll look into this.
- The tests appear to be failing due to a control data mismatch. Needs to be investigated.

After that, we should merge this: generally speaking, this PR is in a pretty good shape already with the more common rendering options working as intended.

Allowing users to pass custom options to `av` would be a nice-to-have / new feature anyways -- and a more radical installation instruction rewrite should also rather happen on a separate PR.

jeertmans commented:
Hello @behackl, this would be great if we can finalize this PR in the next days/weeks.

I would like to emphasize a few points that are, in my opinion, important to consider for this PR:

- Fix #3648: Manim somehow produces empty videos, which are silently skipped* when concatenating because FFMPEG's CLI does a lot of cleaning, that PyAV does not do automatically. This is an error that one user of Manim Slides discovered https://github.com/jeertmans/manim-slides/issues/390 (because Manim Slides already uses PyAV) and the easiest work-around I found was to probe each file and remove files without any video stream;
- Implement #3347 by directly writing audio to partial movies files. I know I am biased because this is a feature request of mine, but IMO that would really benefit Manim Slides, and also potentially Manim users that want to manipulate partial movie files :-)

I initially planned to create and use a separate PyAV library, to reduce to code-complexity (see https://github.com/jeertmans/avfilters), but tackling all possible video formats (especially multiple video streams, audio streams, and subtitles) can be very challenging, and is out of the scope of Manim or Manim Slides. Still, I think you may find interesting code, especially for testing if two media are the same, or else (see https://github.com/jeertmans/avfilters/blob/main/tests/utils.py).

Nevertheless, I think this is important to note that testing for equality (or close) between media is very hard, as many encodings will somehow be lossy.

*: Actually, using `ffmpeg` to concat four 1 second files + one empty file (see https://github.com/jeertmans/avfilters/blob/48d6f01bbb755660e5b5493d6b49a5629f228ae0/tests/test_concatenate.py#L67-L84) results in a 3.9 seconds files (the empty file seems to the

Let me know what you think :-)

behackl commented:
Hey @jeertmans!

- Regarding #3648: we've found that these partial movie files with run_time 0 are also causing issues with the current ffmpeg CLI-based implementation. From my point of view, these files are pathological; there should not be any animations being rendered with a `run_time` of 0 seconds. There has been some effort (#3491) to disallow creating these files, and I am open to introducing more restrictions if necessary (I'll check whether the problem reported [in manim-slides](https://github.com/jeertmans/manim-slides/issues/390) persists with #3491).

- As for #3347, it looks to me as if the main problem with an implementation of this functionality -- the requirement of a mechanism that cuts audio of tracks playing across multiple partial movie files and redistributes the pieces correctly -- is unrelated to this PR. In the spirit of getting the av migration done, I'd rather declare this as out of scope for this PR. I'm not against the feature itself though, I can see how it is useful to have the audio in the partial movie files.

Thanks for the pointer to [jeertmans/avfilters](https://github.com/jeertmans/avfilters)! You are, of course, right -- video tests are inherently hard. I'm already more or less satisfied with proper metadata tests (which you provided an initial rewritten version via av of on this branch here already :-)). Moreover, we have our frame comparison tests still in place anyways.

jeertmans commented:
> Hey @jeertmans!

Thanks for you comprehensive reply, let me address the two points separately.

> * Regarding [FFMPEG concat issue - Manim produces empty media and silently ignores warning #3648](https://github.com/ManimCommunity/manim/issues/3648): we've found that these partial movie files with run_time 0 are also causing issues with the current ffmpeg CLI-based implementation. From my point of view, these files are pathological; there should not be any animations being rendered with a `run_time` of 0 seconds. There has been some effort ([Fix animations with zero runtime length to give a useful error instead of a broken pipe #3491](https://github.com/ManimCommunity/manim/pull/3491)) to disallow creating these files, and I am open to introducing more restrictions if necessary (I'll check whether the problem reported [in manim-slides](https://github.com/jeertmans/manim-slides/issues/390) persists with [Fix animations with zero runtime length to give a useful error instead of a broken pipe #3491](https://github.com/ManimCommunity/manim/pull/3491)).

I wasn’t aware of #3491 and this is, to me, a valid solution to #3648. The idea of just that we need to report that error to the user in a meaningful way, and not a libav error :) Which should now be the case here.

> * As for [Add audio in partial_movie_files, not at the end #3347](https://github.com/ManimCommunity/manim/issues/3347), it looks to me as if the main problem with an implementation of this functionality -- the requirement of a mechanism that cuts audio of tracks playing across multiple partial movie files and redistributes the pieces correctly -- is unrelated to this PR. In the spirit of getting the av migration done, I'd rather declare this as out of scope for this PR. I'm not against the feature itself though, I can see how it is useful to have the audio in the partial movie files.

Sure, adding audio to each segment might not be an easy feature, and probably needs to rewrite the actual thing. If you think this is not part of this PR’s, which I sort of agree, then I don’t mind creating another PR for that feature, once this one is closed (as the audio feature will also use av). Nonetheless, I just wanted to point that so we can keep in mind that you might add an audio stream in the future, and so hopefully we do not write code that is against that :)

jeertmans commented on manim/scene/scene_file_writer.py:
```diff
@@ -634,33 +684,40 @@ def combine_to_movie(self):
+                "metadata": f"comment=Rendered with Manim Community v{__version__}",
+            }
+
+            video_input = av.open(str(movie_file_path))
+            audio_input = av.open(str(sound_file_path))
```
I actually might suggest using context managers instead of `open/close`, so `with av.open(...) as video_input, av.open() as audio_input:`.

jeertmans replied:
The above logic can be applied (almost) everywhere, as the cost of additional indentation level.

behackl replied:
Good call, we've in fact forgotten to close some opened containers. Done in 60abd6e4, but I haven't changed all of them to avoid double-indentations. (It could also be rewritten further to avoid that though.) 👍

behackl commented on manim/utils/commands.py:
```diff
@@ -27,21 +28,25 @@ def capture(command, cwd=None, command_input=None):
+        stream = container.streams.video[0]
+        ctxt = stream.codec_context
+        rate = stream.average_rate
+        if stream.duration is not None:
```
Yes, for .webm (VP9-encoded) files I had to implement this specific workaround.

jeertmans replied:
Ok Nice! I cannot find (only did a quick Google search) information backing up why VP9 would not include that information, but, anyway, I think this might be worth putting a small comment saying that it was observed for some formats, for example webm, that this field is None, so this information is not lost :)

jeertmans replied:
Ok nvm I found a page documenting that: https://trac.ffmpeg.org/wiki/FFprobeTips#Streamduration

behackl commented on manim/utils/commands.py:
```diff
@@ -27,21 +28,25 @@ def capture(command, cwd=None, command_input=None):
+        ctxt = stream.codec_context
+        rate = stream.average_rate
+        if stream.duration is not None:
+            duration = float(stream.duration) * stream.time_base
```
Pydantic is indeed a great idea; work in progress for the general config over at #3473. I'm not convinced it pays off to introduce it here for the metadata. 🤔

behackl commented:
- there is a nice parametrized test for video metadata given different formats and cli-options now. not all formats (e.g., gif) are tested yet.
- the ubuntu runners apparently segfault 😢 
- transparent vp9-encoded files are currently broken as the new test reveals.

regardless, progress is being made here!

behackl commented:
Good news: the pipeline looks already much better.

Still to test:
- `--format=gif`
- Scenes with included audio

Bad news: something weird is going on with transparency in webm-files. It appears that this branch produces partial video files that are actually transparent, but this transparency is lost as soon as the files are recombined to the final output.

Indeed, even though we are using the `yuva420p` pixel format to save the partial movie files, opening such a file with `av` and inspecting the pixel format of the video stream returns 
```
>>> container = av.open("/home/behackl/code/manim/example_scenes/media/videos/bla/720p30/partial_movie_files/StarScene/4117703622_1377611505_3088195676.webm")
>>> container.streams[0].pix_fmt
'yuv420p'
```

I am not quite sure how to salvage this yet.

behackl commented:
With the change in 4c458366, manim correctly produces transparent vp9-encoded (webm) files, verified by inspection -- which, however, I can't test because I can't seem to find a way to extract the first frame with pyav in a way that exposes the alpha channel.

(And conveniently, calling `VideoFrame.to_ndarray(format="rgba")` on a frame extracted from the video causes a segmentation fault...)

jeertmans commented:
> With the change in [4c45836](https://github.com/ManimCommunity/manim/commit/4c45836629a7b0a724a6d6ed3841b80d317201c9), manim correctly produces transparent vp9-encoded (webm) files, verified by inspection -- which, however, I can't test because I can't seem to find a way to extract the first frame with pyav in a way that exposes the alpha channel.
> 
> (And conveniently, calling `VideoFrame.to_ndarray(format="rgba")` on a frame extracted from the video causes a segmentation fault...)

Great work, thanks!! Here is what I found about transparency with PyAV/ffmpeg:
- https://github.com/PyAV-Org/PyAV/issues/500#issuecomment-513621698
- and the first comment in https://www.reddit.com/r/ffmpeg/s/fgdOzMiuVZ
- 

Maybe this helps?

behackl commented:
It did indeed help a little, thanks! But there is still a bit of mystery left. If I extract the first frame using

```py
from av.codec.context import CodecContext
context = CodecContext.create("libvpx-vp9", "r")
packet = next(container.demux(video=0))
first_frame = context.decode(packet)[0].to_ndarray()
```

then I actually get the correct RGBA-values for the pixel in the upper left corner -- however, the red pixel in the bottom right corner now yields (255, 252, 0, 0) (in ARGB) instead of (255, 255, 0, 0); the red-value is slightly wrong now. Maybe this is some sort of color space issue. I'll play around with this some more.

jeertmans commented:
> It did indeed help a little, thanks! But there is still a bit of mystery left. If I extract the first frame using
> 
> ```python
> from av.codec.context import CodecContext
> context = CodecContext.create("libvpx-vp9", "r")
> packet = next(container.demux(video=0))
> first_frame = context.decode(packet)[0].to_ndarray()
> ```
> 
> then I actually get the correct RGBA-values for the pixel in the upper left corner -- however, the red pixel in the bottom right corner now yields (255, 252, 0, 0) (in ARGB) instead of (255, 255, 0, 0); the red-value is slightly wrong now. Maybe this is some sort of color space issue. I'll play around with this some more.

Nice! Yes I also observed that it is hard to get the same pixel values, probably due to color space or maybe lossy encodings. This is why is use assert_all_close :)

behackl commented:
- gifs are now covered by the parametrized test as well,
- for mp4 + mov audio works (just tested locally, not yet covered by the test)
- TODO: fix rendering the audio track for `--format=webm`

JasonGrace2282 commented:
Just a reminder to update with main and change the CI for MacOS-13 to MacOS-latest

behackl commented:
I did change the pipeline to `macos-latest` again, but it seems that there are now problems with pango. We might want to delay having a working `macos-latest` pipeline until after this PR has been merged and deal with this in a follow-up PR.

jeertmans commented:
> * gifs are now covered by the parametrized test as well,
> * for mp4 + mov audio works (just tested locally, not yet covered by the test)
> * TODO: fix rendering the audio track for `--format=webm`

@behackl Can I help you with one or more?

behackl commented:
@jeertmans thanks, I have some time today to work on this today, I'm somewhat optimistic that this can be finished. :-)

Current state of things: In 5fe5bb5 I've produced some code that works in principle:
- We no longer attempt to add an audio stream to gifs,
- Audio for webm files works, but the current implementation (where an `pydub.AudioSegment` is exported to an `opus` file) _unfortunately requires ffmpeg_. I'll investigate whether we can just write the audio files to `.wav` files and reencode them internally using pyav again.

I'd then like to modify the codec test to check for the presence of a non-empty audio track, but that should be comparatively easy as soon as the pipeline runs. Will keep sending updates here.

JasonGrace2282 commented:
Hey, overall amazing job here on this PR! I believe this will make Manim much easier to use for the average user :)
Seeing as the tests passed (including the new ones), I tried to "break" the rendering system and came up with a case where it fails. Shouldn't be too hard to fix :)
Other than that, LGTM! Thanks for your work on this PR :)

## Small Wait time
```py
def construct(self):
    m = ManimBanner()
    self.play(m.expand())
    self.wait(1e-9) # this fails
```

jeertmans commented:
> Hey, overall amazing job here on this PR! I believe this will make Manim much easier to use for the average user :) Seeing as the tests passed (including the new ones), I tried to "break" the rendering system and came up with a case where it fails. Hopefully 🤞 it shouldn't be too hard to fix :) Other than that, LGTM! Thanks for your work on this PR :)
> 
> ## Small Wait time
> ```python
> def construct(self):
>     m = ManimBanner()
>     self.play(m.expand())
>     self.wait(1e-9) # this fails
> ```

Hello, does it fail to render (I.e., an error is raised) or just fails to add the wait time?

behackl commented:
It full on fails. Good find!

<details>
<summary>Stack trace</summary>

```
╭─────────────────────────────── Traceback (most recent call last) ────────────────────────────────╮
│ /Users/behackl/code/manim/manim/cli/render/commands.py:120 in render                             │
│                                                                                                  │
│   117 │   │   │   try:                                                                           │
│   118 │   │   │   │   with tempconfig({}):                                                       │
│   119 │   │   │   │   │   scene = SceneClass()                                                   │
│ ❱ 120 │   │   │   │   │   scene.render()                                                         │
│   121 │   │   │   except Exception:                                                              │
│   122 │   │   │   │   error_console.print_exception()                                            │
│   123 │   │   │   │   sys.exit(1)                                                                │
│                                                                                                  │
│ /Users/behackl/code/manim/manim/scene/scene.py:239 in render                                     │
│                                                                                                  │
│    236 │   │   │   return True                                                                   │
│    237 │   │   self.tear_down()                                                                  │
│    238 │   │   # We have to reset these settings in case of multiple renders.                    │
│ ❱  239 │   │   self.renderer.scene_finished(self)                                                │
│    240 │   │                                                                                     │
│    241 │   │   # Show info only if animations are rendered or to get image                       │
│    242 │   │   if (                                                                              │
│                                                                                                  │
│ /Users/behackl/code/manim/manim/renderer/cairo_renderer.py:268 in scene_finished                 │
│                                                                                                  │
│   265 │   def scene_finished(self, scene):                                                       │
│   266 │   │   # If no animations in scene, render an image instead                               │
│   267 │   │   if self.num_plays:                                                                 │
│ ❱ 268 │   │   │   self.file_writer.finish()                                                      │
│   269 │   │   elif config.write_to_movie:                                                        │
│   270 │   │   │   config.save_last_frame = True                                                  │
│   271 │   │   │   config.write_to_movie = False                                                  │
│                                                                                                  │
│ /Users/behackl/code/manim/manim/scene/scene_file_writer.py:452 in finish                         │
│                                                                                                  │
│   449 │   │   if write_to_movie():                                                               │
│   450 │   │   │   if hasattr(self, "writing_process"):                                           │
│   451 │   │   │   │   self.writing_process.terminate()                                           │
│ ❱ 452 │   │   │   self.combine_to_movie()                                                        │
│   453 │   │   │   if config.save_sections:                                                       │
│   454 │   │   │   │   self.combine_to_section_videos()                                           │
│   455 │   │   │   if config["flush_cache"]:                                                      │
│                                                                                                  │
│ /Users/behackl/code/manim/manim/scene/scene_file_writer.py:671 in combine_to_movie               │
│                                                                                                  │
│   668 │   │   │   return                                                                         │
│   669 │   │                                                                                      │
│   670 │   │   logger.info("Combining to Movie file.")                                            │
│ ❱ 671 │   │   self.combine_files(                                                                │
│   672 │   │   │   partial_movie_files,                                                           │
│   673 │   │   │   movie_file_path,                                                               │
│   674 │   │   │   is_gif_format(),                                                               │
│                                                                                                  │
│ /Users/behackl/code/manim/manim/scene/scene_file_writer.py:636 in combine_files                  │
│                                                                                                  │
│   633 │   │   │   │   output_container.mux(packet)                                               │
│   634 │   │                                                                                      │
│   635 │   │   else:                                                                              │
│ ❱ 636 │   │   │   for packet in partial_movies_input.demux(partial_movies_stream):               │
│   637 │   │   │   │   # We need to skip the "flushing" packets that `demux` generates.           │
│   638 │   │   │   │   if packet.dts is None:                                                     │
│   639 │   │   │   │   │   continue                                                               │
│                                                                                                  │
│ in demux:166                                                                                     │
│                                                                                                  │
│ in av.container.core.Container.err_check:286                                                     │
│                                                                                                  │
│ in av.error.err_check:328                                                                        │
╰──────────────────────────────────────────────────────────────────────────────────────────────────╯
FileNotFoundError: [Errno 2] No such file or directory:
'/Users/behackl/code/manim/example_scenes/media/videos/bug/720p30/partial_movie_files/BugScene/partial_movie_file_list.t
xt'; last error log: [concat] Impossible to open
'file:/Users/behackl/code/manim/example_scenes/media/videos/bug/720p30/partial_movie_files/BugScene/543634251_3305253770
_1995303018.mp4'
```

</details>

jeertmans commented:
@behackl, this is probably because the wait time is shorted than `1/fps`. I haven't looked at this at the moment, but we probably should log some warning (saying "some animation is too short, i.e., smaller than the 1/fps, and will not be part of the final video") and skip it?

behackl commented:
Yes, that's the reason for sure. I need to do some more testing concerning wait and animation times that aren't integer multiples of the frame rate. I'm a bit undecided about what to do in the case above, I see two possible solutions:

- Raise an exception and forbid animations that are too short (which would be consistent with the case where `run_time=0`),
- Emit a warning and round the specified time up to `1/config.frame_rate` (or, in case we have a general problem with non-integer multiples, to the closest integer multiple of `1/config.frame_rate`).

(At this point, I'd also like to at least document my dissatisfaction with the fact that `config.frame_rate` is actually not the frame rate, but the number of frames per second. :cry: )

behackl commented:
I've pushed a fix for resolving the problem with too short `run_time`s: setting a `run_time` that is positive but shorter than the current frame rate (`1/config.frame_rate` ...) emits a warning and rounds the animation run time up to the current frame rate, such that at least one frame is produced.

Ready for re-review!

jeertmans commented:
> I've pushed a fix for resolving the problem with too short `run_time`s: setting a `run_time` that is positive but shorter than the current frame rate (`1/config.frame_rate` ...) emits a warning and rounds the animation run time up to the current frame rate, such that at least one frame is produced.
> 
> Ready for re-review!

To me, that looks like the right solution to adopt! I'll review the code again in my spare time today or later this week :-)

jeertmans commented on manim/animation/composition.py:
```diff
@@ -91,6 +91,17 @@ def begin(self) -> None:
+            1 / config.frame_rate
+        )  # the naming here is unfortunate, config holds fps
+        if self.run_time < frame_rate:
+            logger.warning(
+                f"Original run time of {self} is shorter than current frame "
+                f"rate ({frame_rate} sec) which cannot be rendered. Rendering "
+                "with the shortest possible duration instead."
```
```suggestion
            logger.warning(
                f"Original run time of {self} is shorter than current frame "
                f"rate (1 frame every {frame_rate:.2f} sec.) which cannot be rendered. "
                "Rendering with the shortest possible duration instead."
```

jeertmans commented on manim/scene/scene_file_writer.py:
```diff
@@ -5,12 +5,14 @@
 __all__ = ["SceneFileWriter"]
 
 import json
+import logging
```
Is it still unused? If so, shouldn't we remove this?

jeertmans reviewed (changes requested):
So I re-reviewed the complete code. See my suggested changes :-)

I have a small question: shouldn't we recompile the translations as we modified some files in the docs? (I found references to FFmpeg in `.pot` files.)
Otherwise, this looks great to me!

behackl commented:
Thanks for the review, I've addressed all your raised points!

Yes, the translation workflow is currently not really taken care of. I'll make it a priority to look into how the mechanism with crowdin worked exactly so that we can properly regenerate the translations. Not on this PR though (and there will be a follow-up dealing with documentation anyways). 👍

jeertmans commented:
> Thanks for the review, I've addressed all your raised points!
> 
> Yes, the translation workflow is currently not really taken care of. I'll make it a priority to look into how the mechanism with crowdin worked exactly so that we can properly regenerate the translations. Not on this PR though (and there will be a follow-up dealing with documentation anyways). 👍

Nice! Then I'm happy to approve this PR! :-)

JasonGrace2282 reviewed (changes requested):
Based on the discussion on discord, `Wait` overrides `Animation.begin`, and as such the warning + rounding does not occur. Other than that, it looks good to me!

behackl commented:
The problem is now actually fixed via 878debf. Essentially, this was because the `Wait` animation overrides `Animation.begin`, which meant the check was now running.

I've moved it (and actually, the check for `run_time <= 0` too) to the function that determines the run time of animations in the context of `Scene.play`; this should have made it more robust. I've also added some more tests to really make sure this has been dealt with.

JasonGrace2282 commented on manim/scene/scene.py:
```diff
@@ -1030,12 +1030,28 @@ def get_run_time(self, animations: list[Animation]):
+                    f"Original run time of {animation} is shorter than current frame "
+                    f"rate (1 frame every {frame_rate:.2f} sec.) which cannot be rendered. "
+                    "Rendering with the shortest possible duration instead."
+                )
```
With the current way this is implemented, doing something like
```py
self.play(
    Succession(
        Create(Circle()),
        Create(Star()),
        lag_ratio=1,
    ),
    run_time=1e-3
)
```
Results in two warnings. It would be nice if this could be fixed, maybe in a follow-up PR?

behackl replied:
Interesting, not sure I can explain this intuitively. Would need to investigate; sounds like a job for a follow-up PR. 🧐

JasonGrace2282 commented on manim/scene/scene_file_writer.py:
```diff
@@ -567,37 +557,92 @@ def combine_files(
+        if not includes_sound:
+            av_options["an"] = "1"
+
+        partial_movies_input = av.open(
```
Is `av.open` not used as a context manager here to save indentation?

behackl replied:
Yes, exactly.

JasonGrace2282 reviewed (approved):
Looks good to me! Thanks for your work here!
